server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -82,103 +82,112 @@ class ObjectService(
             expiresAtString =>
               for {
                 ea <- Either.catchNonFatal(expiresAtString.toLong).fold[IO[Option[Long]]](_ => logger.warn(s"Failed to convert $expiresAtString to epoch millis") *> IO.pure(None), l => IO.pure(Some(l)))
                 instant = ea.map(Instant.ofEpochMilli)
               } yield instant
           }
-        } yield Some(GcsMetadata(lastLockedBy, expiresAt, crc32c, generation))
+        } yield Some(GcsMetadata(localPath, lastLockedBy, expiresAt, crc32c, generation))
       case _ => IO.pure(None)
     }
   } yield res
 
   def checkMetadata(req: GetMetadataRequest): IO[MetadataResponse] =
     for {
       traceId <- IO(TraceId(randomUUID()))
-      storageLinks <- storageLinksCache.get
-      baseDirectory <- IO.fromEither(getLocalBaseDirectory(req.localObjectPath).leftMap(s => BadRequestException(s)))
-      storageLink = storageLinks.find(x => x._1.path == baseDirectory)
-      res <- retrieveMetadata(storageLink, traceId, req.localObjectPath)
-    } yield res
-
-  private def retrieveMetadata(storageLink: Option[(LocalBasePath, StorageLink)], traceId: TraceId, localPath: java.nio.file.Path): IO[MetadataResponse] =
-    storageLink.fold[IO[MetadataResponse]](IO.raiseError(StorageLinkNotFoundException(s"No storage link found for ${localPath}"))) { pair =>
-      val isSafeMode = pair._1 match {
-        case LocalBaseDirectoryPath(_) => false
-        case LocalSafeBaseDirectoryPath(_) => true
-      }
-      val sl = pair._2
-      for {
-        fullBlobName <- IO.fromEither(getFullBlobName(localPath, sl.cloudStorageDirectory.blobPath).leftMap(s => BadRequestException(s)))
-        metadata <- retrieveGcsMetadata(sl.cloudStorageDirectory.bucketName, fullBlobName, traceId)
-        result <- metadata match {
-          case None =>
-            IO(MetadataResponse.RemoteNotFound(isSafeMode, sl))
-          case Some(GcsMetadata(lastLockedBy, expiresAt, crc32c, generation)) =>
-            val localAbsolutePath = config.workingDirectory.resolve(localPath)
-            for {
-              fileBody <- io.file.readAll[IO](localAbsolutePath, blockingEc, 4096).compile.to[Array]
-              calculatedCrc32c = Crc32c.calculateCrc32c(fileBody)
-              syncStatus = if (calculatedCrc32c == crc32c) SyncStatus.Live else SyncStatus.Desynchronized
-              currentTime <- timer.clock.realTime(TimeUnit.MILLISECONDS)
-              lastLock <- expiresAt.flatTraverse[IO, WorkbenchEmail] { ea =>
-                if (currentTime > ea.toEpochMilli)
-                  IO.pure(none[WorkbenchEmail])
-                else
-                  IO.pure(lastLockedBy)
-              }
-            } yield MetadataResponse.EditMode(syncStatus, lastLock, isSafeMode, generation, sl)
+      storageLink <- findStorageLink(req.localObjectPath)
+      res <- storageLink.fold[IO[MetadataResponse]](IO.raiseError(StorageLinkNotFoundException(s"No storage link found for ${req.localObjectPath}"))) { pair =>
```

Comment:
this method is heavily nested (i think at one point there are 6 layers of complexity, but counting it is a bit difficult in git). considering refactoring some logic to a helper function for readability.